h/views/api_auth.py

```diff
@@ -2,11 +2,100 @@
 
 from __future__ import unicode_literals
 
+from pyramid.httpexceptions import HTTPBadRequest, HTTPFound
+from pyramid.view import view_config, view_defaults
+from pyramid import security
+
 from h.exceptions import OAuthTokenError
 from h.util.view import cors_json_view
 from h.util.datetime import utc_iso8601
 
 
+@view_defaults(route_name='oauth_authorize')
+class OAuthAuthorizeController(object):
+
+    def __init__(self, request):
+        self.request = request
+        self.oauth_svc = self.request.find_service(name='oauth')
+        self.user_svc = self.request.find_service(name='user')
+
+    @view_config(request_method='GET',
+                 renderer='h:templates/oauth/authorize.html.jinja2')
+    def get(self):
+        """
+        Check the user's authentication status and present the authorization
+        page.
+        """
+        self._check_params()
+
+        if self.request.authenticated_userid is None:
+            raise HTTPFound(self.request.route_url('login', _query={
+                              'next': self.request.url}))
+
+        params = self.request.params
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        return {'username': user.username,
+                'client_name': 'Hypothesis',
+                'client_id': params['client_id'],
+                'response_type': params['response_type'],
+                'response_mode': params['response_mode'],
+                'state': params.get('state')}
+
+    @view_config(request_method='POST',
+                 renderer='h:templates/oauth/post_authorize.html.jinja2',
+                 effective_principals=security.Authenticated)
+    def post(self):
+        """
+        Process an authentication request and return an auth code to the client.
+
+        Depending on the "response_mode" parameter the auth code will be
+        delivered either via a redirect or via a `postMessage` call to the
+        opening window.
+        """
+        authclient = self._check_params()
+
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        # Create an "authorization code" for the response.
+        # This is in fact just a JWT grant token since auth codes have not yet
```

Comment:
My plan was that use of JWTs would be temporary, until auth codes are implemented, and that both here and in the client there would be comments making that clear. eg: https://github.com/hypothesis/client/blob/3f0416fed84e494f11e11a70f050bcab4a7e7e32/src/sidebar/oauth-auth.js#L107

Perhaps catch up about this on Slack?